test/models/user_test.rb

```diff
@@ -512,6 +512,16 @@ def setup
     assert_equal old_encrypted_password, u.encrypted_password, "Changed password"
   end
 
+  test "#grant_permission" do
+    user = create(:user)
+    application = create(:application)
+    supported_permission = create(:supported_permission, application:)
+
+    user.grant_permission(supported_permission)
+
+    assert user.has_permission?(supported_permission)
+  end
```

Comment:
I think it might be worth adding another test here, to exercise the branch of code where the user already has the supported permission (i.e. the `first` bit of `first_or_create`).

Reply:
I expanded the test in [62b4a86](https://github.com/alphagov/signon/pull/2442/commits/62b4a86517943f4e0e41aaf583f423ac4934e773) to test this branch.